### DISCUSSION TITLE
Remove leader keymap, replace it with keypad

### DISCUSSION THREAD

**Discussion by DogLooksGood** [meow-edit/meow#190] [Ideas]
Branch: `replace-leader-with-keypad`.

The idea is: merging leader and keypad. In vanilla Emacs or distribution like doom, `C-c letter` is used for user defined key bindings. 
The keymap at `C-c` is called `mode-specific-map`, which is not independent on major/minor modes, but as an entry point for customization.

From *elisp > Tips > Key Binding Conventions* in the manual.
> Sequences consisting of ‘C-c’ and a letter (either upper or lower case; ASCII or non-ASCII) are reserved for users; they are the *only* sequences reserved for users.

# Changes

1. `SPC` will enter KEYPAD state, then there are three types of translation. 
- h/c/x(keys in variable) start with `C-<key>`
- m/g start with `M-` or `C-M-`, modifier only 
- other keys, start with `C-c <key>`

2. a new command`meow-keypad`. Use this command to switch to keypad-mode. Bound to `SPC` in NORMAL/MOTION keymap.
3. `meow-leader-keymap` will be removed.
4. `meow-leader-define-key` will define keys in `mode-specific-map`. 

# Example configuration
```
(meow-leader-define-key
 '("a b" . ibuffer))

;; the same as

(define-key mode-specific-map
  (kbd "a b") 'ibuffer)

;; the same as

(global-set-key (kbd "C-c a b") 'ibuffer)
```

# Benefits

1. `C-c` as secondary leader, works in INSERT state.
2. Should work seamlessly with any configuration that use `C-c` as leader, e.g. doom emacs without evil.
3. Popup help after `SPC`?

# Breaking things

1. A binding to `C-c letter` with `meow-leader-define-key` may cause error. But we can prevent by ignoring key binding like this.
2. Users are forced to use `letter`, since `C-c non-letter` may be occupied.

# TODOs
- [x] refactor keypad
- [x] update document
- [x] update tutor
- [x] improve helper functions

**Comment by DogLooksGood** [comment]
Now it's possible to have the popup after single SPC.

![image](https://user-images.githubusercontent.com/11796018/150391163-658b7ccd-7224-4610-907b-45e87cfd8322.png)

**Reply by DogLooksGood** [reply]
A simpler graph for switching state. Having a single command for keypad state switching looks better.
![image](https://user-images.githubusercontent.com/11796018/150600471-184fb19c-5c05-4433-af19-40c2f67ae72f.png)

**Comment by DogLooksGood** [comment]
Merged into master.

**Comment by nowislewis** [comment]
I found that there are many C-c keys in some modes(especially orgmode), it seems dirty when mixing leader-define keys with these default keys. So I wonder if there is a way to set a prefix by users, for example, C-c m is equivalent to the space under motion/normal entering the keypad, which can ensure the cleanness of the leader-definr-key. 

This retains nearly all of the current advantages, including the ability to use C-c m instead of space to enter the leader-define-key in insert mode. And it also supports other non-letter prefixes, for example “space tab”

**Reply by DogLooksGood** [reply]
Yes, I found it too. I'd like to make the "leader" keymap customizable. By default, it's `mode-specific-map`, but can be set to `doom-leader-map`, user defined keymap, etc.

**Reply by mclearc** [reply]
Just to clarify -- are you saying this is already customizable, or that you'd like to make it so? I've tried changing the leader map in `meow-keymap-alist` but that doesn't seem to have any effect. In any case it would be really great if the user had more control over what key is the leader key and which map is used (and of course this all seems very easily doable in a way that doesn't end up with meow's bindings colliding/shadowing with other vanilla keybinds).

**Reply by DogLooksGood** [reply]
I haven't make it, but I will. It's definitely better to support a user specified leader  map. I'm just still thinking about some details.I want to make sure it's the last change on keypad.

**Reply by DogLooksGood** [reply]
I have an update to support replacing leader keymap, by adding this before defining keys in leader.
```
(add-to-list 'meow-keymap-alist (cons 'leader your-leader-keymap))
```

**Reply by mclearc** [reply]
This works really well for me -- thanks very much1

**Comment by DogLooksGood** [comment]
It looks like we can have input simulating by setting parent to `key-translation-map`. I think I found a way to reimplement keypad in a simpler way, and solve the issues mentioned in #64.

**Reply by DogLooksGood** [reply]
Since there are very few benefits, I'd like to stick with current implementation at the moment.
But PRs are welcomed.

**Comment by repelliuss** [comment]
I would like to achieve this:

1. Show the leader keybindings in keypad
2. Show only the bindings with alpha-numerical letters (org-mode uses symbols, are there other exceptions too?).
3. Leader keybindings should override the `mode-specific-map` keybinding, thus only accessible explicitly, like `SPC c <key>`
4. Your description of behavior should persist

Is this possible? I guess I can achieve this by binding `meow-keypad-leader-dispatch` to something else other than `C-c` and doing my bindings to that prefix with the exception of 3. Though, `meow-keypad-leader-dispatch` seems not to work.

It would be also good to have an option not to have c/h/x/m/g ability.

**Reply by DogLooksGood** [reply]
Except 2, looks like you are looking for a normal leader? A leader is just a keymap, how about have a keymap, its parent is the mode-specific-map, and bind the keymap to `SPC` in NORMAL  and MOTION state.

**Comment by lakkiy** [comment]
Should this be updated?
https://github.com/meow-edit/meow/blob/master/TUTORIAL.org#keypad
